File: app/src/main/java/com/geridea/trentastico/network/request/RequestSender.java
package com.geridea.trentastico.network.request;


/*
 * Created with ♥ by Slava on 29/03/2017.
 */

import com.geridea.trentastico.Config;

import java.io.IOException;
import java.util.Timer;
import java.util.TimerTask;
import java.util.Vector;

import okhttp3.Call;
import okhttp3.Callback;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.Response;
import okhttp3.ResponseBody;

public class RequestSender {

    private final OkHttpClient client = new OkHttpClient();
    private final Vector<Call> callsInProgress = new Vector<>();

    private final Timer timeoutWaiter = new Timer();

    public void processRequest(IRequest requestToSend) {
        waitForDebuggingIfNeeded();

        Request request = new Request.Builder()
                .url(requestToSend.getURL())
                .build();

        requestToSend.notifyOnBeforeSend();

        Call call = client.newCall(request);
        call.enqueue(new RequestCallback(requestToSend));


        callsInProgress.add(call);
    }

    private void waitForDebuggingIfNeeded() {
        if (Config.DEBUG_MODE && Config.PRE_LOADING_WAITING_TIME_MS != 0) {
            try {
                Thread.sleep(Config.PRE_LOADING_WAITING_TIME_MS);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
    }

    public void processRequestAfterTimeout(final IRequest requestToSend) {
        timeoutWaiter.schedule(new TimerTask() {
            @Override
            public void run() {
                processRequest(requestToSend);
            }
        }, Config.WAITING_TIME_AFTER_A_REQUEST_FAILED);
    }

    private class RequestCallback implements Callback {

        private IRequest request;

        RequestCallback(IRequest request) {
            this.request = request;
        }

        @Override public void onFailure(Call call, IOException e) {
            callsInProgress.remove(call);
            request.notifyFailure(e, RequestSender.this);
        }

        @Override public void onResponse(Call call, Response response) throws IOException {
            if (response.isSuccessful()) {
                try {
                    ResponseBody responseBody = response.body();
                    String responseStr = responseBody.string();
                    responseBody.close();

                    request.manageResponse(responseStr, RequestSender.this);
                } catch (Exception e) {
                    request.notifyFailure(e, RequestSender.this);
                }
            } else {
                request.notifyResponseUnsuccessful(response.code(), RequestSender.this);
            }
            callsInProgress.remove(call);
        }
    }

}

File: app/src/main/java/com/geridea/trentastico/network/request/RequestSender.kt
package com.geridea.trentastico.network.request


/*
 * Created with ♥ by Slava on 29/03/2017.
 */

import com.geridea.trentastico.Config
import com.geridea.trentastico.utils.IS_IN_DEBUG_MODE
import okhttp3.*
import java.io.IOException
import java.util.*

class RequestSender {

    private val client = OkHttpClient()

    private val timeoutWaiter = Timer()

    fun processRequest(requestToSend: IRequest) {

        waitForDebuggingIfNeeded()
        requestToSend.notifyOnBeforeSend()

        val call = client.newCall(buildRequest(requestToSend))
        call.enqueue(RequestCallback(requestToSend))
    }

    fun processAsyncRequest(requestToSend: IRequest){
        waitForDebuggingIfNeeded()
        requestToSend.notifyOnBeforeSend()

        try {
            val call = client.newCall(buildRequest(requestToSend))
            call.execute().use { response ->
                onProcessResponseWithRequest(response, requestToSend)
            }
        } catch (e: Exception) {
            requestToSend.notifyNetworkProblem(e, this)
        }
    }

    private fun onProcessResponseWithRequest(response: Response, request: IRequest) {
        if (response.isSuccessful) {
            try {
                val responseStr = response.body()!!.use { it.string() }

                request.manageResponse(responseStr, this@RequestSender)
            } catch (e: Exception) {
                if (IS_IN_DEBUG_MODE) {
                    e.printStackTrace()
                }
                request.notifyResponseProcessingFailure(e, this@RequestSender)
            }
        } else {
            request.notifyNetworkProblem(
                    ResponseUnsuccessfulException(response.code()),
                    this@RequestSender
            )
        }
    }

    private fun buildRequest(requestToSend: IRequest): Request {
        val builder = Request.Builder()
        builder.url(requestToSend.url)

        //If we have something to post, we'll get it here
        val formBodyToSend = requestToSend.formToSend
        if (formBodyToSend != null) {
            builder.post(formBodyToSend)
        }

        return builder.build()
    }

    private fun waitForDebuggingIfNeeded() {
        if (IS_IN_DEBUG_MODE && Config.PRE_LOADING_WAITING_TIME_MS != 0) {
            try {
                Thread.sleep(Config.PRE_LOADING_WAITING_TIME_MS.toLong())
            } catch (e: InterruptedException) {
                e.printStackTrace()
            }

        }
    }

    fun processRequestAfterTimeout(requestToSend: IRequest) = timeoutWaiter.schedule(object : TimerTask() {
        override fun run() = processRequest(requestToSend)
    }, Config.WAITING_TIME_AFTER_A_REQUEST_FAILED.toLong())

    private inner class RequestCallback internal constructor(private val request: IRequest) : Callback {

        override fun onFailure(call: Call, e: IOException) {
            request.notifyNetworkProblem(e, this@RequestSender)
        }

        @Throws(IOException::class)
        override fun onResponse(call: Call, response: Response) {
            onProcessResponseWithRequest(response, request)
        }
    }

}
